src/libclipper/include/clipper/threadpool.hpp

```diff
@@ -300,6 +300,37 @@ class ModelQueueThreadPool : public ThreadPool {
     }
   }
 
+  bool delete_queue(VersionedModelId vm, int replica_id) {
+    boost::unique_lock<boost::shared_mutex> l(queues_mutex_);
+    size_t queue_id = get_queue_id(vm, replica_id);
+    auto queue = queues_.find(queue_id);
+    if (queue == queues_.end() || !queue->second.is_valid()) {
+      log_error_formatted(LOGGING_TAG_THREADPOOL,
+                          "Work queue does not exist for model {}, replica {}",
+                          vm.serialize(), std::to_string(replica_id));
+      return false;
+    } else {
+      queue->second.invalidate();
+      l.unlock();
+
+      auto thread = threads_.find(queue_id);
+      if (thread->second.joinable()) {
+        thread->second.join();
```

Comment:
Did you have a success to join with the terminated thread in this code?
I think that it causes a failure because TaskExecutionThread is waiting for a signal of queue_not_empty_condition_ conditional variable in the TaskExecutor.

Reply:
The relevant queue is first invalidated, which signals all waiters on the condition variable, so this shouldn't be an issue; the thread should finish any current running task and terminate.